apps/emqx_foreman/src/emqx_foreman.erl

```diff
@@ -0,0 +1,837 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_foreman).
+
+%%========================================================================================
+%% @doc This process implements leader election on top of `global' to distribute
+%% "resources" to members of its `pg' group.
+%%
+%% Upon initialization, each process must be given a `pg' scope and name, which should
+%% only be shared by processes of the same kind (e.g.: the same connector type and name).
+%%
+%% This process can be in one of 3 states: `leader', `follower', `candidate'.
+%%
+%% The process starts in the `candidate' state and attempts to elect itself by claiming a
+%% `global' name based on the input scope and name.  If it manages to claim it, it's the
+%% `leader'.  Otherwise, it monitors the leader pid and enters the `follower' state.  A
+%% `follower' will transition back to a `candidate' if it deems that the leader is dead.
+%%
+%% If a name clash is detected by `global' (e.g.: after cliques in a netsplit merge back),
+%% the leaders in each clique will shut themselves down to signal their followers to
+%% trigger a new election.
+%%
+%% Once a leader is elected, it'll evaluate an allocation of resources based on the
+%% current `pg' group members and an input callback function, and then "stage" these
+%% allocations by casting a message to each member if it's a new allocation (i.e.:
+%% different from previously computed).  Each process has a generation id to track
+%% evolving configurations, which is bumped on new allocations and included in allocation
+%% messages.
+%%
+%% If a `follower' receives an assignment tagged with an older generation id, it sends the
```

Comment:
how can this happen? 
a stale leader recovered from network partition ?

Reply:
a leader might die, be restarted fast enough and grab the global name again